### PR TITLE
AS-585: first pass at delta layer module

### PR DESCRIPTION
**Pull request by davidangb**
For context: https://lucid.app/lucidchart/invitations/accept/22b09176-eaca-44bc-a469-61b8de4f3766 and https://cloud.google.com/solutions/streaming-data-from-cloud-storage-into-bigquery-using-cloud-functions

This PR creates a new Terraform module for Delta Layer, which creates three buckets,  two SAs, and applies permissions for those SAs to the buckets. Eventually this module will do more, but we're aiming to develop/deploy incrementally.

The corresponding PR in terraform-ap-deployments: broadinstitute/terraform-ap-deployments#227

### CODE REVIEW

File: deltalayer/sa.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Service account for the "streamer" cloud function
+resource "google_service_account" "sa_streamer" {
+  count = var.enable && contains(["preview_shared"], var.env_type) ? 1 : 0
```
Comment (mikeflinn) [inline deltalayer/sa.tf]:
```suggestion
  count = var.enable && contains(["preview_shared"], var.env_type) ? 1 : 0
```
```suggestion
  count = var.enable ? 1 : 0
```
the `contains(["preview_shared"]...` part of this conditional isn't needed here. That is a remnant from an older tf pattern we are moving away from

File: deltalayer/provider.tf [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  alias = "target"
+}
+
+provider "google" {
+  alias = "dns"
+}
+
+provider "google-beta" {
+  alias = "target"
+}
```
Comment (mikeflinn) [inline deltalayer/provider.tf]: I would delete this entirely. This tf module isn't creating anything that requires google-beta

**Comment by davidangb** [comment]
@mflinn-broad @zarsky-broad this is ready for review! See broadinstitute/terraform-ap-deployments#227 as the calling PR, which includes the atlantis plan.

File: deltalayer/buckets.tf [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  provider = google.target
+  project  = var.google_project
+  location = var.bucket_location
+}
+
+resource "google_storage_bucket_iam_binding" "success-bucket-sa-binding-sourcewriter" {
+  bucket   = google_storage_bucket.source-bucket.name
+  provider = google.target
+  role     = "roles/storage.objectCreator"
+  members  = ["serviceAccount:${var.sourcewriter_sa_email}"]
```
Comment (zarsky-broad) [inline deltalayer/buckets.tf:15]: Looks like these support `member`, which doesn't have to be an array: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_iam#member/members

File: deltalayer/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+output "sa_streamer_id" {
+  value       = google_service_account.sa_streamer[0].account_id
+  description = "Streamer SA ID"
+}
+
+output "sa_filemover_id" {
+  value       = google_service_account.sa_filemover[0].account_id
+  description = "File-mover SA ID"
+}
```
Comment (zarsky-broad) [inline deltalayer/outputs.tf]: nit: newline
```suggestion
}

```

File: deltalayer/buckets.tf (hunk 3 above)
Comment (zarsky-broad) [reply deltalayer/buckets.tf:15]: Alternately, you could simplify by using one [`iam_policy`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/iam_policy) with multiple `binding`s in it, and then [applying it to the bucket](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_iam#google_storage_bucket_iam_policy)

**Review by zarsky-broad — APPROVED** [review]
Minor simplification thoughts, but basically looks solid to me.

**Review by mikeflinn — APPROVED** [review]
lgtm,  I don't have anything to add that Isaac didn't mention already. 

Just a friendly reminder to make sure to make sure this will actually `atlantis apply` successfully from your ap-deployments pr before merging this

**Comment by davidangb** [comment]
@zarsky-broad @mflinn-broad I've updated to use `google_iam_policy` - please re-review if you choose to, I will interpret silence as approval! (and thank you)

**Review by mikeflinn — APPROVED** [review]
👍🏻

**Pull request merged**